### ISSUE TITLE
Stored XSS in budget name

### ISSUE THREAD

**Issue by dayn1ne** [firefly-iii/firefly-iii#2335]
**Description**
Current version of Firefly III Version 4.7.17 is vulnerable to stored XSS due to lack of filtration of user-supplied data in budget name. Malicious attacker can create specially crafted link, which contains javascript code and assign it to budget name. User who add transactions with malicious budget, can view page with that transaction on tag summary page (http://firefly.host/tags/show/$tag_number$), malicious javascript code will be executed.

**Steps to reproduce**
I create a small video for reproduce steps
[XSS Steps.zip](https://github.com/firefly-iii/firefly-iii/files/3392669/XSS.Steps.zip)

**PoC**
```
POST /budgets/store HTTP/1.1
Host: localhost
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Content-Type: application/x-www-form-urlencoded
Content-Length: 114
Cookie: firefly_session=<Your session here>;
DNT: 1
Connection: close
Upgrade-Insecure-Requests: 1

_token=<Your token here>=1&name=<script>alert("XSS PoC")</script>
```
<img width="945" alt="Screenshot 2019-07-15 at 17 18 30" src="https://user-images.githubusercontent.com/1087660/61223084-a62ae100-a724-11e9-8f62-cc535f38b49b.png">

**Comment by JC5** [comment]
Thanks, nice find. I'll verify and push an update as soon as I can. You're mixing up tags and budgets by the way, is this correct?

**Comment by JC5** [comment]
Sorry, I see what you mean.

**Comment by JC5** [comment]
I have pushed 4.7.17.1 that should fix the bug and a similar one in category names. The docker builds are running now. In my haste, I completely forgot to credit you which I will do in my release post and on reddit, after you confirm the issue is gone.

I've reopened the issue as GitHub has automatically closed it.

**Comment by dayn1ne** [comment]
> I have pushed 4.7.17.1 that should fix the bug and a similar one in category names. The docker builds are running now. In my haste, I completely forgot to credit you which I will do in my release post and on reddit, after you confirm the issue is gone.
> 
> I've reopened the issue as GitHub has automatically closed it.

I check commit - 45ddb64 and see only changes to versions in some files and change titles in other.

But anyway i got source code from https://github.com/firefly-iii/firefly-iii/releases/tag/4.7.17.1 and install it for checking.

Vulnerability still exist

I uploaded the video to the dropbox - (https://www.dropbox.com/s/gdr87bh4pat5ly5/xss_2.mov?dl=0)

**Comment by JC5** [comment]
You're right. I am an idiot. I pushed the correct fix in 4.7.17.2. Can you confirm?

**Comment by dayn1ne** [comment]
> You're right. I am an idiot. I pushed the correct fix in 4.7.17.2. Can you confirm?

Confirm, i cannot reproduce XSS

Thank you for such quick fix and for this wonderful budget app.
I will continue search for security bugs and will inform you if found them

**Comment by JC5** [comment]
No problem, please do. You know where to find me!
